### ISSUE TITLE
Reconciler and GlusterFS: E1129 an empty namespace may not be set when a resource name is provided

### ISSUE THREAD

**Issue by TerraTech** [kubernetes/kubernetes#37625]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6", GitCommit:"e569a27d02001e343cb68086bc06d47804f62af6", GitTreeState:"clean", BuildDate:"2016-11-12T05:22:15Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.6+coreos.0", GitCommit:"ec2b52fabadf824a42b66b6729fe4cff2c62af8c", GitTreeState:"clean", BuildDate:"2016-11-14T19:42:00Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
CoreOS-1185.3.0 (baremetal)
hyperkube-v1.4.6_coreos.0

**What happened**:
I am seeing this repeating in my logs every 3 minutes
```
Nov 29 08:37:06 core-minion05 kubelet-wrapper[1148]: E1129 13:37:06.881298    1148 glusterfs.go:109] glusterfs: failed to get endpoints gtfs-pool-0005[an empty namespace may not be set when a resource name is provided]
Nov 29 08:37:06 core-minion05 kubelet-wrapper[1148]: E1129 13:37:06.881346    1148 reconciler.go:432] Could not construct volume information: MountVolume.NewMounter failed for volume "kubernetes.io/glusterfs/d1142b29-b634-11e6-9b60-0018f3531b4f-gtfs-pool-0005" (spec.Name: "gtfs-pool-0005") pod "d1142b29-b634-11e6-9b60-0018f3531b4f" (UID: "d1142b29-b634-11e6-9b60-0018f3531b4f") with: an empty namespace may not be set when a resource name is provided
```
```sh
$ kubectl get pv gtfs-pool-0005 -oyaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: 2016-11-06T10:32:32Z
  labels:
    type: gtfs
  name: gtfs-pool-0005
  resourceVersion: "34998685"
  selfLink: /api/v1/persistentvolumes/gtfs-pool-0005
  uid: 537f898f-a40c-11e6-85f3-0018f3531b4f
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: kube-grafana
    namespace: kube-system
    resourceVersion: "34998683"
    uid: 2f1be42a-b634-11e6-9b60-0018f3531b4f
  glusterfs:
    endpoints: glusterfs-cluster
    path: gtfs-pool-0005
  persistentVolumeReclaimPolicy: Retain
status:
  phase: Bound
```
```sh
$ kubectl get pvc kube-grafana -oyaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: 2016-11-29T13:03:12Z
  labels:
    k8s-app: influxGrafana
  name: kube-grafana
  namespace: kube-system
  resourceVersion: "34998687"
  selfLink: /api/v1/namespaces/kube-system/persistentvolumeclaims/kube-grafana
  uid: 2f1be42a-b634-11e6-9b60-0018f3531b4f
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  volumeName: gtfs-pool-0005
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  phase: Bound
```
Given the E1129 error, are Persistent Volumes supposed to be namespaced?

**Comment by olmoser** [comment]
I'm running into the same issue, with the difference that in our setup, there is no PV (just the glusterfs service and endpoint. This is what they look like:

**glusterfs endpoint** (deployed in all available namespaces)
```
Name:		glusterfs-cluster
Namespace:	default
Labels:		<none>
Subsets:
  Addresses:		10.2.56.10
  NotReadyAddresses:	<none>
  Ports:
    Name	Port	Protocol
    ----	----	--------
    <unset>	1	TCP

No events.
```

**glusterfs service** (deployed in all namespaces as well):

```
Name:			glusterfs-cluster
Namespace:		default
Labels:			<none>
Selector:		<none>
Type:			ClusterIP
IP:			10.106.112.86
Port:			<unset>	1/TCP
Endpoints:		10.2.56.10:1
Session Affinity:	None
No events.
```

**Environment**
RHEL 7.3 Bare Metal Setup with kubeadm
glusterfs 3.9.0

```
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.1-dirty", GitCommit:"82450d03cb057bab0950214ef122b67c83fb11df", GitTreeState:"dirty", BuildDate:"2017-01-11T09:40:43Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.0", GitCommit:"58b7c16a52c03e4a849874602be42ee71afdcab1", GitTreeState:"clean", BuildDate:"2016-12-12T23:31:15Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by hiscal2015** [comment]
@TerraTech @olmoser I have the same issue, did you find the root cause?

**Comment by humblec** [comment]
@TerraTech @olmoser @hiscal2015 can u give some more context abt the issue. Are these PVCs statically provisioned or dynamically provisioned? If these are statically provisioned, can you please create an endpoint and service in the namespace where you run the pod which make use of subjected PVClaim ?

**Comment by iT2afL0rd** [comment]
I am also run into this error.

kubenetes cluster version: 
```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

glusterfs service and ep:
```
# kubectl get svc glusterfs-cluster
NAME                CLUSTER-IP       EXTERNAL-IP   PORT(S)   AGE
glusterfs-cluster   10.254.150.248   <none>        1/TCP     127d

[root@ipattern-core26 bricks]# kubectl get ep glusterfs-cluster
NAME                ENDPOINTS                                      AGE
glusterfs-cluster   10.40.*.11:1,10.40.*.12:1,10.40.*.13:1   127d
```

The pv and pvc are statically provisioned:
```
# kubectl get pv mlp-redis -oyaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: 2017-03-15T09:00:04Z
  name: mlp-redis
  resourceVersion: "17176953"
  selfLink: /api/v1/persistentvolumesmlp-redis
  uid: c76bee9b-095d-11e7-9b71-44a842169a4d
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 30Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: mlp-redis
    namespace: default
    resourceVersion: "17176951"
    uid: cb29d6d0-095d-11e7-9b71-44a842169a4d
  glusterfs:
    endpoints: glusterfs-cluster
    path: mlp-redis
  persistentVolumeReclaimPolicy: Recycle
status:
  phase: Bound
```

pvc
```
# kubectl get pvc mlp-redis -oyaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: 2017-03-15T09:00:10Z
  name: mlp-redis
  namespace: default
  resourceVersion: "17176955"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/mlp-redis
  uid: cb29d6d0-095d-11e7-9b71-44a842169a4d
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 30Gi
  volumeName: mlp-redis
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 30Gi
  phase: Bound
```

And I got no errors in glusterfs volume logs, the gfs volume works normally both inside and outside of the cluster.

**Comment by cjyar** [comment]
I'm also seeing this error. PVCs which were working suddenly stopped working after some of the nodes rebooted last night to install OS updates. I'm seeing the same error: `an empty namespace may not be set when a resource name is provided`

Kubernetes 1.6.4+coreos.0
CoreOS 1409.2.0

**Comment by cjyar** [comment]
I fixed this on my cluster. I had 2 glusterfs pods that were not running, even though `kubectl get pods` showed them as `Running`. Restarting the two glusterfs pods (by using `docker rm` on the node) made this seemingly irrelevant error go away, and everything started working.

- Affected pods are on nodes that have rebooted recently.
- Heketi, on a recently-rebooted node, is running fine.
- Node uptimes:
    - 1, 2, 3, 7, 9, and 10 rebooted last night.
    - 4, 5, 6, and 8 did not reboot last night.
- GlusterFS is running on nodes 1, 2, 3, 5, 6, 8, and 9.
- GlusterFS on node 1 says 2 and 3 are disconnected. Node 5 also claims 2 and 3 are disconnected.
- Can't exec into glusterfs pod on nodes 2 or 3: `error: unable to upgrade connection: container not found ("glusterfs")`
- `kubectl describe pod` says:
```
Name:           glusterfs-t7lzw
Namespace:      kube-system
Node:           cse-mt-dev-002/123.45.67.89
Start Time:     Fri, 28 Apr 2017 11:36:33 -0600
Labels:         app=glusterfs
                glusterfs-node=pod
                pod-template-generation=1
Annotations:    kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"DaemonSet","namespace":"kube-system","name":"glusterfs","uid":"bf3d029e-292c-11e7-9dc8-003048f9b3b8","apiV...
                kubernetes.io/psp=privileged
Status:         Running
IP:             123.45.67.89
Controllers:    DaemonSet/glusterfs
Containers:
  glusterfs:
    Container ID:       docker://5d83cffe7fb16b03f2934d2976aac16d8035f06b1cdeb41966856c9d513e520e
    Image:              gluster/gluster-centos:latest
    Image ID:           docker-pullable://gluster/gluster-centos@sha256:e3e2881af497bbd76e4d3de90a4359d8167aa8410db2c66196f0b99df6067cb2
    Port:
    State:              Running
      Started:          Wed, 31 May 2017 01:05:44 -0600
    Last State:         Terminated
      Reason:           Completed
      Exit Code:        0
      Started:          Mon, 01 Jan 0001 00:00:00 +0000
      Finished:         Wed, 31 May 2017 01:00:15 -0600
    Ready:              True
    Restart Count:      2
    Liveness:           exec [/bin/bash -c systemctl status glusterd.service] delay=60s timeout=3s period=10s #success=1 #failure=3
    Readiness:          exec [/bin/bash -c systemctl status glusterd.service] delay=60s timeout=3s period=10s #success=1 #failure=3
    Environment:        <none>
    Mounts:
      /dev from glusterfs-dev (rw)
      /etc/glusterfs from glusterfs-etc (rw)
      /etc/ssl from glusterfs-ssl (ro)
      /run from glusterfs-run (rw)
      /run/lvm from glusterfs-lvm (rw)
      /sys/fs/cgroup from glusterfs-cgroup (ro)
      /var/lib/glusterd from glusterfs-config (rw)
      /var/lib/heketi from glusterfs-heketi (rw)
      /var/lib/misc/glusterfsd from glusterfs-misc (rw)
      /var/log/glusterfs from glusterfs-logs (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-2n160 (ro)
Conditions:
  Type          Status
  Initialized   True
  Ready         False
  PodScheduled  True
Volumes:
  glusterfs-heketi:
    Type:       HostPath (bare host directory volume)
    Path:       /var/lib/heketi
  glusterfs-run:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
  glusterfs-lvm:
    Type:       HostPath (bare host directory volume)
    Path:       /run/lvm
  glusterfs-etc:
    Type:       HostPath (bare host directory volume)
    Path:       /etc/glusterfs
  glusterfs-logs:
    Type:       HostPath (bare host directory volume)
    Path:       /var/log/glusterfs
  glusterfs-config:
    Type:       HostPath (bare host directory volume)
    Path:       /var/lib/glusterd
  glusterfs-dev:
    Type:       HostPath (bare host directory volume)
    Path:       /dev
  glusterfs-misc:
    Type:       HostPath (bare host directory volume)
    Path:       /var/lib/misc/glusterfsd
  glusterfs-cgroup:
    Type:       HostPath (bare host directory volume)
    Path:       /sys/fs/cgroup
  glusterfs-ssl:
    Type:       HostPath (bare host directory volume)
    Path:       /etc/ssl
  default-token-2n160:
    Type:       Secret (a volume populated by a Secret)
    SecretName: default-token-2n160
    Optional:   false
QoS Class:      BestEffort
Node-Selectors: storagenode=glusterfs
Tolerations:    <none>
Events:
  FirstSeen     LastSeen        Count   From                    SubObjectPath   Type            Reason          Message
  ---------     --------        -----   ----                    -------------   --------        ------          -------
  9h            9s              2618    kubelet, cse-mt-dev-002                 Warning         FailedSync      Error syncing pod, skipping: rpc error: code = 2 desc = unable to inspect docker image "sha256:d261a90b2feca43cda077490b6fddf03cf756b4ae1859c546de34648cb3ad48f" while inspecting docker container "2d414a91ccc43144debece19bd05fd28209da3bd13e81768ee3d7e1624ef806d": no such image: "sha256:d261a90b2feca43cda077490b6fddf03cf756b4ae1859c546de34648cb3ad48f"
```
- docker on node 2 says the GlusterFS pod isn't running:
```
CONTAINER ID        IMAGE                                                                                                                       COMMAND                  CREATED              STATUS                        PORTS               NAMES
2d414a91ccc4        d261a90b2fec                                                                                                                "/usr/sbin/init"         3 weeks ago          Exited (0) 3 weeks ago                            k8s_glusterfs_glusterfs-t7lzw_kube-system_38831c24-2c39-11e7-ad5d-002590d74e1e_1
```
- `docker rm 2d414a91ccc4` causes the container to restart.
    - Gluster on other nodes says node 2 is now online.
    - Pods which were failing to start have recovered and are working normally now.

In summary, these are the problems:
- Pods were listed as `Running` but were, in fact, not running.
- The error message `an empty namespace may not be set when a resource name is provided` actually means, "Can't talk to one of your glusterfs pods."

**Comment by mliker** [comment]
Hi, I am also experiencing this, however, I am running openshift origin 1.5.1 (kubernetes master v1.5.2+43a9be40). I am using dynamic provisioning. Although I haven't noticed anything being broken, the log lines are unsettling and repeat periodically:

```
Jul 12 09:37:51 n1 origin-node: E0712 09:37:51.300684    3230 glusterfs.go:133] glusterfs: failed to get endpoints pvc-709d4948-66cf-11e7-bbda-901b0ecbcd7e[an empty namespace may not be set when a resource name is provided]
Jul 12 09:37:51 n1 origin-node: E0712 09:37:51.300728    3230 reconciler.go:437] Could not construct volume information: MountVolume.NewMounter failed for volume "kubernetes.io/glusterfs/eeba66a3-66cf-11e7-9a8c-901b0e9623b6-pvc-709d4948-66cf-11e7-bbda-901b0ecbcd7e" (spec.Name: "pvc-709d4948-66cf-11e7-bbda-901b0ecbcd7e") pod "eeba66a3-66cf-11e7-9a8c-901b0e9623b6" (UID: "eeba66a3-66cf-11e7-9a8c-901b0e9623b6") with: an empty namespace may not be set when a resource name is provided
```

storage class def:
```apiVersion: storage.k8s.io/v1beta1
kind: StorageClass
metadata:
  name: gluster
  annotations:
    storageclass.beta.kubernetes.io/is-default-class: "true"
provisioner: kubernetes.io/glusterfs
parameters:
  resturl: "http://10.1.0.1:8080"
  restuser: "admin"
  secretName: "heketi-secret"
  secretNamespace: "default"
```

service and endpoints were created by heketi:

```root@n1 ~ # oc get svc glusterfs-dynamic-docker-registry -o yaml
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: 2017-07-12T06:58:04Z
  labels:
    gluster.kubernetes.io/provisioned-for-pvc: docker-registry
  name: glusterfs-dynamic-docker-registry
  namespace: default
  resourceVersion: "775068"
  selfLink: /api/v1/namespaces/default/services/glusterfs-dynamic-docker-registry
  uid: 73bc1b8f-66cf-11e7-9a8c-901b0e9623b6
spec:
  clusterIP: 172.30.164.75
  ports:
  - port: 1
    protocol: TCP
    targetPort: 1
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

root@n1 ~ # oc get ep glusterfs-dynamic-docker-registry -o yaml
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: 2017-07-12T06:58:04Z
  labels:
    gluster.kubernetes.io/provisioned-for-pvc: docker-registry
  name: glusterfs-dynamic-docker-registry
  namespace: default
  resourceVersion: "775066"
  selfLink: /api/v1/namespaces/default/endpoints/glusterfs-dynamic-docker-registry
  uid: 73b2c64b-66cf-11e7-9a8c-901b0e9623b6
subsets:
- addresses:
  - ip: 10.0.0.1
  - ip: 10.1.0.1
  - ip: 10.2.0.1
  ports:
  - port: 1
    protocol: TCP
```

Dynamic pvc claim in dc of docker-registry:
```root@n1 ~ # oc get dc docker-registry -o yaml
apiVersion: v1
kind: DeploymentConfig
metadata:
  creationTimestamp: 2017-06-28T15:31:05Z
  generation: 10
  labels:
    docker-registry: default
  name: docker-registry
  namespace: default
  resourceVersion: "775452"
  selfLink: /oapi/v1/namespaces/default/deploymentconfigs/docker-registry
  uid: ccc70878-5c16-11e7-94b6-901b0ecbcd7e
spec:
  replicas: 1
  selector:
    docker-registry: default
  strategy:
    activeDeadlineSeconds: 21600
    resources: {}
    rollingParams:
      intervalSeconds: 1
      maxSurge: 25%
      maxUnavailable: 25%
      timeoutSeconds: 600
      updatePeriodSeconds: 1
    type: Rolling
  template:
    metadata:
      creationTimestamp: null
      labels:
        docker-registry: default
    spec:
      containers:
      - env:
        - name: REGISTRY_HTTP_ADDR
          value: :5000
        - name: REGISTRY_HTTP_NET
          value: tcp
        - name: REGISTRY_HTTP_SECRET
          value: 7EU0c/vM6euNvpwEpAVPiaLCfMv7iHryAVr/dJcY5Vg=
        - name: REGISTRY_MIDDLEWARE_REPOSITORY_OPENSHIFT_ENFORCEQUOTA
          value: "false"
        - name: REGISTRY_HTTP_TLS_KEY
          value: /etc/secrets/registry.key
        - name: REGISTRY_HTTP_TLS_CERTIFICATE
          value: /etc/secrets/registry.crt
        image: openshift/origin-docker-registry:v1.5.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 5000
            scheme: HTTPS
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: registry
        ports:
        - containerPort: 5000
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /healthz
            port: 5000
            scheme: HTTPS
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          requests:
            cpu: 100m
            memory: 256Mi
        securityContext:
          privileged: false
        terminationMessagePath: /dev/termination-log
        volumeMounts:
        - mountPath: /etc/secrets
          name: registry-certificates
        - mountPath: /registry
          name: volume-v8gtq
      dnsPolicy: ClusterFirst
      nodeSelector:
        region: infra
      restartPolicy: Always
      securityContext: {}
      serviceAccount: registry
      serviceAccountName: registry
      terminationGracePeriodSeconds: 30
      volumes:
      - name: registry-certificates
        secret:
          defaultMode: 420
          secretName: registry-certificates
      - name: volume-v8gtq
        persistentVolumeClaim:
          claimName: docker-registry
  test: false
  triggers:
  - type: ConfigChange
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2017-07-05T09:33:24Z
    lastUpdateTime: 2017-07-05T09:33:24Z
    message: Deployment config has minimum availability.
    status: "True"
    type: Available
  - lastTransitionTime: 2017-07-12T07:01:42Z
    lastUpdateTime: 2017-07-12T07:01:44Z
    message: replication controller "docker-registry-5" successfully rolled out
    reason: NewReplicationControllerAvailable
    status: "True"
    type: Progressing
  details:
    causes:
    - type: ConfigChange
    message: config change
  latestVersion: 5
  observedGeneration: 10
  readyReplicas: 1
  replicas: 1
  unavailableReplicas: 0
  updatedReplicas: 1
```

**Comment by hlubek** [comment]
Actually I think there are 2 issues in `reconciler.go` when reconstructing the volume spec and executing `plugin.NewMounter`:

1. No namespace can be reconstructed from the pod and volume dir
2. The endpoints for GlusterFS (when not using a PVC) are set to the volume name, which is clearly wrong (`glusterfs.go:203`)

Both seem to be a conceptual issue where the data cannot be reconstructed only from the directory names.

**Comment by humblec** [comment]
Looking into this.

**Comment by jsafrane** [comment]
I think that iSCSI is not that widely used and nobody has stumbled upon this. The code is obviously wrong.

**Comment by rootfs** [comment]
fix is in https://github.com/kubernetes/kubernetes/pull/54264
@mtanino
